Display dishes and dish details

flamesoft opened the pull request:
PR story: 
- https://www.pivotaltracker.com/story/show/121661019
- https://www.pivotaltracker.com/story/show/121660703

Activities:
- Displays price for dishes
- Display name for every dish
- By clicking a dish name, the visitor is directed to the dish details page
- Display an image for a dish
- Display calories
- Display allergy information

Notes: 
- Menu is not shown on dish list because menu model is not ready
- Ingredients are not shown because the priority is decreased

Commit:
Add dishes and related files
WIP- add tests,
need to do - finish the pending tests

Commit:
Remove comments

Commit:
Add a test for factorygirl

Commit:
Show image on dish details page

Commit:
Adjust style for dish details page

Commit:
Show image on the dishes list

Commit:
WIP- add acceptance test for dishes

Commit:
Merge branch 'development' into dish

Commit:
Remove dish_controller_spec.rb because it can be duplicate for acceptance test

Commit:
Remove unfinished test
it is connected with menu so the tests will be added after Menu model is done

tochman commented:
@flamesoft  The Create, New, Edit, Delete methods in the controller are obsolete. Comment them out for now or delete them. Also remove the scaffolded comments the generator added to the controller.

flamesoft commented:
Ok. If I comment out the "Create, new, edit, delete" methods, the functions are not working for CRUD. Shall I move them to other places?

Commit:
Comment CRUD methods in dish_controller.rb
These methods are not included in the story
They will be added for a new story

tochman commented:
We need acceptance tests for displaying the list of dishes as well as showing a single dish. I can't see test like that in this PR. The unit tests are okay. I would get rid of the request spec. ( controller spec is also not needed imo, but I'm okay with keeping it.). I also see a lot of unused views. Probably generated by scaffold. They are not needed atm. Feel free to keep them but be aware of them. ( perhaps make a comment about them in the PR description)

Commit:
Remove request spec

Commit:
Add test for showing dishes and prices

Commit:
Add test for display dish details

Commit:
Add test for showing the dish detail
Remove Show link
Make dish name to be clickable and link to the dish detail page

tochman commented on spec/models/dish_spec.rb:
```diff
@@ -0,0 +1,17 @@
+  describe 'DB table' do
+    it { is_expected.to have_db_column :id }
+    it { is_expected.to have_db_column :name }
+    it { is_expected.to have_db_column :alergyInfo }
```
So, the naming standard for attributes ic snake_case :-( Sorry, I missed that in my previous review. (also, it's `allergy` with 2 `l's`)

```
allergy_info
```

tochman commented on spec/requests/dishes_spec.rb:
```diff
@@ -0,0 +1,10 @@
+      expect(response).to have_http_status(200)
+    end
+  end
+end
```
Please get rid of this spec.

tochman commented on spec/routing/dishes_routing_spec.rb:
```diff
@@ -0,0 +1,39 @@
+  describe "routing" do
+
+    it "routes to #index" do
+      expect(:get => "/dishes").to route_to("dishes#index")
```
Plz try to use new hash syntax. (I guess this is auto generated code, right?) I would still update it.

Commit:
Delete dishes_routing_spec

Commit:
Rename alergyInfo to allery_info

tochman commented on db/migrate/20160623133301_change_column_name.rb:
```diff
@@ -0,0 +1,5 @@
+class ChangeColumnName < ActiveRecord::Migration
+  def change
+    rename_column :dishes, :allergy_info, :allergy_info
```
Am I missing something here or are we renaming the column to the same name?

tochman replied:
Since we are creating a new object, I'm worried the test is not actually routing the user to the right show page.

flamesoft replied:
I made replacement one extra time. I will fix this.

tochman commented on features/step_definitions/common_steps.rb:
```diff
@@ -8,6 +8,12 @@
+      visit dishes_path
+    when "dish detail page" then
+      dish = FactoryGirl.create(:dish, name: 'Dimsun', price: '90',
+                                allergy_info: 'soy', calories: '120')
```
Shouldn't this object be created in the background? Why is this created here?

flamesoft replied:
Yes. I did it at the beginning.  But then I had the problem - how can a dish be identified in the system? This is the only way I found.

tochman replied:
I know that. It's not a trivial task. That's why I wanted you to spend some time with @diraulo

tochman commented on app/views/dishes/index.html.haml:
```diff
@@ -0,0 +1,31 @@
+    %tbody
+      - @dishes.each do |dish|
+        %tr
+          %td= link_to dish.name, dish
```
Is this link actually pointing to the dish show page?  I would think that the markup should be:

```
link_to dish.name, dish_path(dish)
```

flamesoft replied:
It works with the test of the click on the link. I have also tested it the web page and it works.

tochman replied:
Okay, fair enough. Those helper methods are truly powerful ;-)

tochman commented on app/views/dishes/index.html.haml:
```diff
@@ -0,0 +1,31 @@
+          %td= dish.calories
+          %td= link_to image_tag(dish.img_url)
+          %td= link_to 'Edit', edit_dish_path(dish)
+          %td= link_to 'Destroy', dish, :method => :delete, :data => { :confirm => 'Are you sure?' }
```
Plz use new hash syntax. Across the board.

flamesoft replied:
Thanks. I will change them.

tochman commented on features/dishes.feature:
```diff
@@ -0,0 +1,32 @@
+  Scenario: Show the dish details page
+    Given I am on the "dish list page"
+    And I click on "Dimsun"
+    Then I should be on the "dish detail page"
```
```
Then I should be on the dish detail page for "Dimsun"
```

In the step definition you can find that object by:

```
dish = Dish.find_by(name: name) #or whatever name you give to the arg)
```

then the assertion could look something like:

```
expect(current_path).to eq dish_path(dish)
```

flamesoft replied:
Thanks! This is the solution that I was looking for.

Commit:
Change old column name to be alergyInfo in migration script

Commit:
Change old hash syntax to new hash syntax for index.html.haml

Commit:
Change old hash syntax to new hash syntax for show.html.haml

Commit:
Change test step description for the dish detail display test

tochman commented:
1. The request spec should be removed and
2. The latest code needs to be pulled in

Then I think we can merge this in. Good job. 

There are some lessons learned from this one. I still have some issues with the way views are designed but I would suggest we return to that in a separate story.

flamesoft commented:
@tochman thanks for your comments. I will fix them.

Commit:
Remove the request test

Commit:
Change show details test

tochman commented:
This on still has conflicts (even before the latests PR merge)

flamesoft commented:
There are conflicts with schema.rb, routing.rb and common_steps.rb. I have tried to fix the conflict but it seems that it must be done by another person. @callea2 will help to merge.

Commit:
Solve conflicts

Commit:
Merge branch 'flamesoft-dish' into development

Commit:
Merge branch 'development' of https://github.com/CraftAcademy/slow_food_online into development
